1.6.7: Error during installation "Cannot find module '../lib/main'"

Hi there,

I'm running into trouble when trying to install recognize app version 1.6.7 on Nextcloud 22.1.1:

```
{
  "reqId": "FzTWvN8HCrJwluRHJMmv",
  "level": 3,
  "time": "2021-08-31T15:33:03+02:00",
  "remoteAddr": "2001:a62:421:3401:f480:1ca7:fd7d:9ade",
  "user": "Administrator",
  "app": "no app in context",
  "method": "POST",
  "url": "/index.php/settings/apps/enable",
  "message": "Exception while executing repair step Install dependencies",
  "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.159 Safari/537.36",
  "version": "22.1.1.2",
  "exception": {
    "Exception": "Exception",
    "Message": "Failed to install Tensorflow.js: CPU-linux-3.8.0.tar.gz\nnode-pre-gyp install failed with error: Error: Command failed: node-pre-gyp install --fallback-to-build\n/bin/sh: 1: node-pre-gyp: Permission denied",
    "Code": 0,
    "Trace": [
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/apps/recognize/lib/Migration/InstallDeps.php",
        "line": 81,
        "function": "runTfjsInstall",
        "class": "OCA\\Recognize\\Migration\\InstallDeps",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/Repair.php",
        "line": 123,
        "function": "run",
        "class": "OCA\\Recognize\\Migration\\InstallDeps",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/legacy/OC_App.php",
        "line": 1059,
        "function": "run",
        "class": "OC\\Repair",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/Installer.php",
        "line": 166,
        "function": "executeRepairSteps",
        "class": "OC_App",
        "type": "::"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/apps/settings/lib/Controller/AppSettingsController.php",
        "line": 447,
        "function": "installApp",
        "class": "OC\\Installer",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/AppFramework/Http/Dispatcher.php",
        "line": 217,
        "function": "enableApps",
        "class": "OCA\\Settings\\Controller\\AppSettingsController",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/AppFramework/Http/Dispatcher.php",
        "line": 126,
        "function": "executeController",
        "class": "OC\\AppFramework\\Http\\Dispatcher",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/AppFramework/App.php",
        "line": 156,
        "function": "dispatch",
        "class": "OC\\AppFramework\\Http\\Dispatcher",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/Route/Router.php",
        "line": 301,
        "function": "main",
        "class": "OC\\AppFramework\\App",
        "type": "::"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/base.php",
        "line": 1000,
        "function": "match",
        "class": "OC\\Route\\Router",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/index.php",
        "line": 36,
        "function": "handleRequest",
        "class": "OC",
        "type": "::"
      }
    ],
    "File": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/apps/recognize/lib/Migration/InstallDeps.php",
    "Line": 112,
    "CustomMessage": "Exception while executing repair step Install dependencies"
  }
}
```

I also tried to add execute permission during installation with a very ugly while loop, however this results in a different error:

```
{
  "reqId": "4UMOXxLciX0iuCBv3von",
  "level": 3,
  "time": "2021-08-31T15:36:46+02:00",
  "remoteAddr": "2001:a62:421:3401:f480:1ca7:fd7d:9ade",
  "user": "Administrator",
  "app": "no app in context",
  "method": "POST",
  "url": "/index.php/settings/apps/enable",
  "message": "Exception while executing repair step Install dependencies",
  "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.159 Safari/537.36",
  "version": "22.1.1.2",
  "exception": {
    "Exception": "Exception",
    "Message": "Failed to install Tensorflow.js: CPU-linux-3.8.0.tar.gz\nnode-pre-gyp install failed with error: Error: Command failed: node-pre-gyp install --fallback-to-build\ninternal/modules/cjs/loader.js:892\n  throw err;\n  ^\n\nError: Cannot find module '../lib/main'\nRequire stack:\n- /mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/apps/recognize/node_modules/.bin/node-pre-gyp\n    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:889:15)\n    at Function.Module._load (internal/modules/cjs/loader.js:745:27)\n    at Module.require (internal/modules/cjs/loader.js:961:19)\n    at require (internal/modules/cjs/helpers.js:92:18)\n    at Object.<anonymous> (/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/apps/recognize/node_modules/.bin/node-pre-gyp:4:1)\n    at Module._compile (internal/modules/cjs/loader.js:1072:14)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)\n    at Module.load (internal/modules/cjs/loader.js:937:32)\n    at Function.Module._load (internal/modules/cjs/loader.js:778:12)\n    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12) {\n  code: 'MODULE_NOT_FOUND',\n  requireStack: [\n    '/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/apps/recognize/node_modules/.bin/node-pre-gyp'\n  ]\n}",
    "Code": 0,
    "Trace": [
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/apps/recognize/lib/Migration/InstallDeps.php",
        "line": 81,
        "function": "runTfjsInstall",
        "class": "OCA\\Recognize\\Migration\\InstallDeps",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/Repair.php",
        "line": 123,
        "function": "run",
        "class": "OCA\\Recognize\\Migration\\InstallDeps",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/legacy/OC_App.php",
        "line": 1059,
        "function": "run",
        "class": "OC\\Repair",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/Installer.php",
        "line": 157,
        "function": "executeRepairSteps",
        "class": "OC_App",
        "type": "::"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/apps/settings/lib/Controller/AppSettingsController.php",
        "line": 447,
        "function": "installApp",
        "class": "OC\\Installer",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/AppFramework/Http/Dispatcher.php",
        "line": 217,
        "function": "enableApps",
        "class": "OCA\\Settings\\Controller\\AppSettingsController",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/AppFramework/Http/Dispatcher.php",
        "line": 126,
        "function": "executeController",
        "class": "OC\\AppFramework\\Http\\Dispatcher",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/AppFramework/App.php",
        "line": 156,
        "function": "dispatch",
        "class": "OC\\AppFramework\\Http\\Dispatcher",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/private/Route/Router.php",
        "line": 301,
        "function": "main",
        "class": "OC\\AppFramework\\App",
        "type": "::"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/lib/base.php",
        "line": 1000,
        "function": "match",
        "class": "OC\\Route\\Router",
        "type": "->"
      },
      {
        "file": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/index.php",
        "line": 36,
        "function": "handleRequest",
        "class": "OC",
        "type": "::"
      }
    ],
    "File": "/mnt/cowssd/web/apps/nextcloud/instances/prod/nextcloud-21.0.0/apps/recognize/lib/Migration/InstallDeps.php",
    "Line": 112,
    "CustomMessage": "Exception while executing repair step Install dependencies"
  }
}
```

Any idea?

--- a/lib/Migration/InstallDeps.php
+++ b/lib/Migration/InstallDeps.php
@@ -78,6 +78,8 @@ class InstallDeps implements IRepairStep
         // Write the app config
         $this->config->setAppValue('recognize', 'node_binary', $binaryPath);
 
+        $this->setBinariesPermissions();
+
         $this->runTfjsInstall($binaryPath);
     }
 
@@ -127,5 +129,18 @@ class InstallDeps implements IRepairStep
         $tar = new TAR($file);
         $tar->extractFile('node-'.$version.'-linux-'.$arch.'/bin/node', $this->binaryDir.'/node');
         return $this->binaryDir.'/node';
+    }
+
+    /**
+     * try to fix binaries permissions issues
+     */
+    protected function setBinariesPermissions()
+    {
+        $iterator = new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->nodeModulesBinaryDir));
+        foreach($iterator as $item) {
+            if (chmod(realpath($item->getPathname()), 0755) === FALSE) {
+                throw new \Exception('Error when setting node_modules/.bin/* permissions');
+            }
+        }
     }
 }

